src/Components/Components/src/Rendering/Renderer.cs

```diff
@@ -461,18 +461,95 @@ private void ProcessRenderQueue()
+            {
+                // The display update was cancelled (maybe due to a timeout on the components server-side case or due
+                // to the renderer being disposed)
+                return Task.CompletedTask;
+            }
+            if (updateDisplayTask.IsFaulted)
+            {
+                // The display update failed so we don't care any more about running on render completed
+                // fallbacks as the entire rendering process is going to be torn down.
+                HandleException(updateDisplayTask.Exception);
+                return Task.CompletedTask;
+            }
+
+            if (!updateDisplayTask.IsCompleted)
+            {
+                var updatedComponentsId = new int[updatedComponents.Count];
+                var updatedComponentsArray = updatedComponents.Array;
+                for (int i = 0; i < updatedComponentsId.Length; i++)
+                {
+                    updatedComponentsId[i] = updatedComponentsArray[i].ComponentId;
+                }
+
+                return InvokeAfterUpdateDisplayCompleted(updateDisplayTask, updatedComponentsId);
+            }
+
             List<Task> batch = null;
             var array = updatedComponents.Array;
             for (var i = 0; i < updatedComponents.Count; i++)
             {
                 var componentState = GetOptionalComponentState(array[i].ComponentId);
                 if (componentState != null)
+                {
+                    ProcessComponent(componentState, ref batch);
+                }
+            }
+
+            return batch != null ?
+                Task.WhenAll(batch) :
+                Task.CompletedTask;
+
+            async Task InvokeAfterUpdateDisplayCompleted(Task updateDisplayTask, int[] updatedComponents)
+            {
+                try
+                {
+                    await updateDisplayTask;
+                }
+                catch when (updateDisplayTask.IsCanceled)
+                {
+                    return;
+                }
+                catch when (updateDisplayTask.IsFaulted)
+                {
+                    HandleException(updateDisplayTask.Exception);
+                    return;
+                }
+
+                List<Task> batch = null;
+                var array = updatedComponents;
+                for (var i = 0; i < updatedComponents.Length; i++)
+                {
+                    var componentState = GetOptionalComponentState(array[i]);
+                    if (componentState != null)
+                    {
+                        ProcessComponent(componentState, ref batch);
+                    }
+                }
+
+                var result = batch != null ?
+                    Task.WhenAll(batch) :
+                    Task.CompletedTask;
+
+                await result;
+            }
+
+            void ProcessComponent(ComponentState state, ref List<Task> batch)
+            {
+                if (state != null)
```

Comment:
All the calls to `ProcessComponent` already verify that `state` is nonnull. We could either remove the null check from here or from the call sites.

Reply:
done